akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1088,43 +1086,68 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
 
     private def sendEvent(event: T, promise: Promise[Done]): Promise[Done] = {
       onAsyncInput(event, OptionVal.Some(promise))
-      currentState.get() match {
-        case Completed ⇒ failPromiseOnComplete(promise)
-        case _         ⇒ promise
-      }
+      if (stopped) failPromiseOnComplete(promise)
+      else promise
     }
 
     // external call
     override def invokeWithFeedback(event: T): Future[Done] = {
       val promise: Promise[Done] = Promise[Done]()
-      promise.future.andThen { case _ ⇒ waitingForProcessing.remove(promise) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
-      waitingForProcessing.add(promise)
+      promise.future.onComplete(_ ⇒ onFeedbackCompleted(promise))(ExecutionContexts.sameThreadExecutionContext)
+
+      @tailrec
+      def addToWaiting(): Unit = {
+        val previous = asyncCallbacksInProgress.get()
+        if (previous != null) {
+          val updated = previous + (this -> (previous(this) + promise))
+          if (!asyncCallbacksInProgress.compareAndSet(previous, updated)) addToWaiting()
+        }
```

Comment:
What happens if `previous == null`? It seems that it runs into the `if (stopped) promise.failure(new StreamDetachedException())` in `invokeWithPromise`? Maybe add a comment or move that handling here.